### PR TITLE
[WIP] Use pre-estimated PSDs in inspiral like optimal_snr.

### PR DESCRIPTION
**Pull request by bhooshan-gadre**
We can use precomputed PSD with inspiral now. 
This should be useful when we run injections. This way, no need to estimate PSD again during inspiral and injections will not correct the PSDs.
This needs to be tested… carefully.

### CODE REVIEW

File: pycbc/psd/__init__.py [hunk 1]
```diff
@@ -522,26 +522,43 @@ def associate_psds_to_segments(opt, fd_segments, gwstrain, flen, delta_f, flow,
...
-            if psd_seg.intersects(inp_seg):
-                curr_overlap = abs(inp_seg & psd_seg)
-                if curr_overlap > psd_overlap:
-                    psd_overlap = curr_overlap
-                    best_psd = psd
-        if best_psd is None:
-            err_msg = "No PSDs found intersecting segment!"
-            raise ValueError(err_msg)
-        fd_segment.psd = best_psd
+    if opt.use_precomputed_psd and opt.precomputed_psd_file:
```
Comment (spxiwh) [inline pycbc/psd/__init__.py:525]: It feels like we only need `opt.precomputed_psd_file` and if this is given we assume that it will be used. (so remove the `--use-precomputed-psd` option).

Comment (bhooshan-gadre) [reply pycbc/psd/__init__.py:525]: Yes, I agree.

**Comment by bhooshan-gadre** [comment]
@spxiwh 
I am currently computing and checking results with this functionality. Hence WIP.

Apart from O4 injection runs, I am also using it to check the possibility of using ET null stream PSD for ET searches in my other work.

File: requirements.txt [hunk 2]
```diff
@@ -19,6 +19,7 @@ gwdatafind>=1.1.3
 
 # Requirements for full pegasus env
 pegasus-wms.api >= 5.0.3
+GitPython
```
Comment (spxiwh) [inline requirements.txt:22]: Why is this needed? The pegasus API does not require it, and we are not trying to install the full pegasus toolkit (ie `pegasus-plan` and friends) here.

Comment (bhooshan-gadre) [reply requirements.txt:22]: I know that this is not needed on CIT but somehow while generating the workflow, I have been asked to install the GitPython package on lho,  llo and hawk when I  built my know env:
```
virtualenv -p /cvmfs/oasis.opensciencegrid.org/ligo/sw/conda/envs/igwn-py39/bin/python env

source env/bin/activate

cd env

git clone https://github.com/gwastro/pycbc.git

cd pycbc
pip install -r requirements.txt
pip install -r companion.txt
pip install .

pip install -r requirements-igwn.txt
pip install .[igwn]
```

**Comment by bhooshan-gadre** [comment]
Now this works. Yet to take a very careful look at the output.

File: bin/pycbc_inspiral [hunk 3]
```diff
@@ -201,9 +201,22 @@ parser.add_argument("--multiprocessing-nprocesses", type=int,
...
+# generated in the workflow.
+# psd.add_argument('--psd-files', metavar='FILE', nargs='*',
+#                         help='Instead of time-independent PSDs, use time-varying '
+#                         'PSDs from the given HDF5 files and pick the appropriate '
+#                         'PSD for each injection')
+
+parser.add_argument("--precomputed-psd-file", metavar='FILE', type=str,
+                        help='Instead of time-independent PSDs, use time-varying '
+                        'PSDs from the given HDF5 files and pick the appropriate '
+                        'PSD for each injection')
```
Comment (GarethCabournDavies) [inline bin/pycbc_inspiral:219]: The argument only takes one option, but the help mentions multiple PSD files.

I think it would be best not to mention injections - I cant think of alternative use-cases, but there might be one!

Comment (GarethCabournDavies) [reply bin/pycbc_inspiral:219]: In fact it may be worth adding this to the `insert_psd_option_group` as this may be useful elsewhere, though I dont know the code well enough to say thats definitely possible

Comment (bhooshan-gadre) [reply bin/pycbc_inspiral]: Done

File: requirements.txt (hunk 2 above)
Comment (spxiwh) [reply requirements.txt:22]: In any case, while I would like to understand this, let's not include it in this PR as it's not related to the change being proposed.

Comment (GarethCabournDavies) [reply requirements.txt:22]: Without wanting to derail the PR (@bhooshan-gadre could you add an issue?) I've seen this when attempting to run workflows at CIT as well

**Comment by GarethCabournDavies** [comment]
Would this functionality allow us to analyse more tightly-spaced injections? So far as I remember, the "injection spacing" discussions we had recently were mostly focussed on PSD bias, but I don't remember if there was any other factors at play

**Comment by bhooshan-gadre** [comment]
> Would this functionality allow us to analyse more tightly-spaced injections? So far as I remember, the "injection spacing" discussions we had recently were mostly focussed on PSD bias, but I don't remember if there was any other factors at play
@GarethCabournDavies Yes, that is one of the ideas. Also, this will allow one to use ET null-stream PSD or PSD post hierarchically removals and save some cost.

**Comment by spxiwh** [comment]
@bhooshan-gadre This one has been open for a while. Can it be cleaned up and merged?
